bpf/bpf_host.c

```diff
@@ -1697,6 +1676,39 @@ int cil_to_host(struct __ctx_buff *ctx)
 	 * to mark as PACKET_OTHERHOST and drop.
 	 */
 	ctx_change_type(ctx, PACKET_HOST);
+#if !defined(TUNNEL_MODE)
+	/* Since v1.18 Cilium performs IPsec encryption at the native device,
+	 * before the packet leaves the host.
+	 *
+	 * A special case exists for L7 egress proxy packets when native routing
+	 * mode is enabled.
+	 *
+	 * Because L7 egress proxy packets are generated in the host-namespace
+	 * and generated packets MUST adjust their MTU for ESP encapsulation
+	 * an IP route MTU adjustment exists for L7 egress proxy packets.
+	 *
+	 * When the L7 egress proxy generates packets an 'ip rule' in the host
+	 * namespace routes these packets into table 2005 which has a route
+	 * toward 'cilium_host' and adjusts the MTU correctly for ESP encap.
+	 *
+	 * When 'cil_from_host@cilium_host' is reached the skb's mark is zeroed
+	 * and the packet is pushed toward 'cil_to_host@cilium_net'.
+	 *
+	 * If we simply let this packet drop to the stack, an iptables rule
+	 * exists which will mark the packet with 0x200 and trigger a local
+	 * delivery as part of L7 Proxy TPROXY mechanism.
+	 *
+	 * This iptables rule is ignored by the mark MARK_MAGIC_PROXY_TO_WORLD.
+	 * Technically, it is also ignored by MARK_MAGIC_ENCRYPT but reusing
+	 * this mark breaks further processing as its used in the XFRM subsystem.
+	 *
+	 * Therefore, if the packet's mark is zero, indicating it was forwarded
+	 * from 'cilium_host', mark the packet with MARK_MAGIC_PROXY_TO_WORLD
```

Comment:
It's been discussed months ago and I really should have brought it up earlier, but, given wireguard routes proxied traffic directly to native devices (on native routing), what if ipsec follows the same approach? This means changing the 2005 routing table from `default dev cilium_host` to ... whatever there are in the "main" routing table, with MTU adjustments.

(actually it's unclear to me why wireguard + proxy never saw mtu issue? I should have a look at it)

then, we might not need this "smelly" code?

Reply:
@jschwinger233 I do not want to get into the scenario where we are constantly trying to `mirror` "main" routing table changes into a shadow "2005" table. The Cilium code has multiple sources of routes, along with endpoint routes being a configuration. I find it a harder sell to say "mirror main into 2005" then to simply send back into `cilium_host`.

Reply:
Agreed with 

> do not want to get into the scenario where we are constantly trying to mirror "main" routing table changes into a shadow "2005" table

otherwise it'll be chaotic when it comes to eni / dynamic routing scenaio.